File: solutions/ruby/complex-numbers/1/complex_numbers.rb
# frozen_string_literal: true

# class ComplexNumber
class ComplexNumber
  attr_reader :real, :imaginary

  def initialize(real, imaginary = 0)
    @real = real.to_f
    @imaginary = imaginary.to_f
  end

  def abs()= Integer.sqrt(@real**2 + @imaginary**2)
  def conjugate = ComplexNumber.new(@real, -@imaginary)
  def exp = ComplexNumber.new(Math.exp(@real) * Math.cos(@imaginary), Math.exp(@real) * Math.sin(@imaginary))
  def ==(other) = (@real - other.real).abs < 1e-15 && (@imaginary - other.imaginary).abs < 1e-15
  def +(other) = ComplexNumber.new(@real + other.real, @imaginary + other.imaginary)
  def -(other) = ComplexNumber.new(@real - other.real, @imaginary - other.imaginary)

  def *(other) = ComplexNumber.new(@real * other.real - @imaginary * other.imaginary,
                                   @imaginary * other.real + @real * other.imaginary)

  def /(other) = ComplexNumber.new(
    (@real * other.real + @imaginary * other.imaginary) / (other.real**2 + other.imaginary**2),
    (@imaginary * other.real - @real * other.imaginary) / (other.real**2 + other.imaginary**2)
  )
end

File: solutions/clojure/complex-numbers/1/src/complex_numbers.clj
(ns complex-numbers
  (:refer-clojure :exclude [abs])
  (:require [clojure.math :as m]))

(defn real [[a _b]] a)
(defn imaginary [[_a b]] b)
(defn abs [[a b]] (m/sqrt (+ (* a a) (* b b))))
(defn conjugate [[a b]] [a, (- b)])
(defn add [[a b] [c d]] [(+ a c), (+ b d)])
(defn sub [[a b] [c d]] [(- a c), (- b d)])
(defn mul [[a b] [c d]] [(- (* a c) (* b d)), (+ (* b c) (* a d))])
(defn div [[a b] [c d]]
  (let [cc (* c c), dd (* d d)]
    [(/ (+ (* a c) (* b d)) (+ cc dd)),
     (/ (- (* b c) (* a d)) (+ cc dd))]))
